PCF-ADR-0004-codeflare-test-strategy.md

```diff
@@ -0,0 +1,392 @@
+
+
+### Test Suite
+
+#### E2E Test Cases
+
+##### Setup
+
+The following steps must be executed once, with a cluster admin role, as prerequisites to running the e2e test cases:
+1. Provision a test Kubernetes cluster (resp. an OpenShift cluster)
+2. Build the CodeFlare operator container image:
+  * Clone the CodeFlare operator source code repository
+  * Checkout the branch to be tested, e.g., the PR feature branch
+  * Build the CodeFlare operator container image at HEAD revision
+  * Push the CodeFlare operator container image into a test container image registry (resp. the OpenShift cluster internal image registry)
+3. Install the CodeFlare stack components into the test cluster:
+  * Configure the CodeFlare operator deployment with Kustomize,  to use the previously built container image
+  * Create the codeflare-system Namespace (resp. Project)
+  * Deploy the CodeFlare operator using Kustomize
+    Note: the installation using OLM is covered as part of the installation / upgrade test cases
+  * Deploy the KubeRay operator
+  * Create a default MCAD resource
+  * Grant the MCAD controller ServiceAccount edit permission on RayCluster resources
+  * Create a default InstaScale resource
+Wait until all the components are ready
+The e2e test cases can now be executed, with a **standard user role**, ideally in parallel, to speed the execution time up and shorten the feedback loop as much as possible.
+> **Note**
+> The test cases should define lean / minimal compute resources, relative to the provisioned cluster, so parallel execution / throughput of the batch jobs submitted via the MCAD scheduling queue is maximized.
+
+##### Submit a Sample PyTorch Job in a managed Ray Cluster
+
+###### Description
+
+Submit a test PyTorch batch job, to a Ray cluster managed by MCAD, in a user tenant, and assert successful completion of the job. Shutdown the Ray cluster, and assert successful freeing of resources.
+
+###### Scenario
+
+1. Create a test Namespace (resp. Project)
+2. Create a test AppWrapper resource with the following specifications: https://github.com/project-codeflare/multi-cluster-app-dispatcher/blob/14d569bec1cd016dd41352e3c026f461d851f480/doc/usage/examples/kuberay/config/aw-raycluster.yaml
+3. Wait until the Ray cluster is ready
+4. Submit the test batch job to the Ray cluster
+5. Wait until the job has completed
+6. Assert the job status is successful
+7. Delete the test AppWrapper resource
+8. Assert all the resources have been successfully freed
+
+##### Submit a Sample PyTorch Job directly via MCAD
+
+Lower priority.
+
+###### Description
+
+Submit a test PyTorch batch job to the MCAD scheduler, in a user tenant, and assert successful completion of the job. Assert successful execution of the job, and clean-up of resources upon the batch job completion.
+
+###### Scenario
```

Comment:
I think the specifics of these cases should live in a design doc or something else. It may be too much technical detail for this ADR. @MichaelClifford what do you think?

Reply:
@anishasthana I agree. I think the flow here is good for the ADR. But the implementation details might change.

Reply:
Dropped technical specifications and replaced them with a general description.